### ISSUE TITLE
may be a bug?

### ISSUE THREAD

**Issue by moonABC3** [LinWeizheDragon/Knowledge-Aware-Graph-Enhanced-GPT-2-for-Dialogue-State-Tracking#1]
Hi, thanks for sharing your code. I found a possible bug. In "graph_modules.py", the original code "aij = nn.functional.softmax(eij * maskEdges - infinityMask, dim=4) ", whether the dimension of softmax is wrong, it should be "aij = nn.functional.softmax(eij * maskEdges - infinityMask , dim=3)" to correspond to the code above "aWx = a1Wx + a2Wx.permute(0, 1, 2, 4, 3)". Looking forward to your reply.

**Comment by LinWeizheDragon** [comment]
Hi, thanks for your question. I was using the code in another project: https://github.com/proroklab/magat_pathplanning

Regarding your question, aWx is B x P x E x N x N. The last two dimensions N x N are always n_nodes x n_nodes with the same dimension and meaning of adjacency matrix.
a1Wx + a2Wx.permute(0, 1, 2, 4, 3) is just a quicker implementation of convolution, so the dimension order has not been changed.
Therefore, softmax should still be used in the last dimension (4), as it makes the aij (node i to j) sum up to 1 for each node i.

**Comment by moonABC3** [comment]
Thanks for your reply. a1Wx is a row vector and a2Wx.permute() is a column vector. For aij, a1Wx + a2Wx = aWx, the ith column of aWx corresponds to node i, not the ith row of aWx corresponds to node i. I think so.

**Comment by LinWeizheDragon** [comment]
The mechanism to pass features from nodes to neighbouring nodes is:
`x = torch.matmul(x, aij)`
And x is B x P x E x G x N,   aij is B x P x E x N x N, the right multiplication leads to B x P x E x G x N.
An example is G = 1, N = 2
x = [0.7, 0.5] (1 x N)
eij = [[1, 1], [1, 1]] (N x N)   # features of nodes are passed to themselves and the neighbours, all weighted by 1
aij  = softmax(eij, dim=-1) = [[0.5, 0.5], [0.5, 0.5]]  # each row is softmaxed
matmul(x, aij) = [0.6, 0.6]  as expected.

**Comment by LinWeizheDragon** [comment]
If you want the answers from whom implemented it, you can create an issue in https://github.com/proroklab/magat_pathplanning. The code is exactly the same.

**Comment by moonABC3** [comment]
OK, thanks again for your answer.
